flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java

```diff
@@ -163,8 +224,19 @@ public T copy(T from, T reuse) {
 
 	@Override
 	public void copy(DataInputView source, DataOutputView target) throws IOException {
-		T value = deserialize(source);
-		serialize(value, target);
+		if (CONCURRENT_ACCESS_CHECK) {
+			enterExclusiveThread();
+		}
+
+		try {
+			T value = deserialize(source);
```

Comment:
Have to point out that after `deserialize()` the checks in copy() are ineffective as the `currentThread` field has already been nulled. In other words, we guard against concurrent access before deserialize(), and within deserialize()&serialize(), but not between deserialize()&serialize() or after serialize().

This isn't a _problem_ as all code is actually covered, but we may want to document that.

Reply:
I think we can simply remove the check around the `copy(DataInputView source, DataOutputView target)` because the respective critical parts (`serialize()` and `deserialize()`) are covered already.

Reply:
An alternative would be to do the following, but I would honestly not do that. This is a heuristic anyways (otherwise we would need to use CAS on the `currentThread` field).

```java
private void enterExclusiveThread() {
	Thread previous = currentThread;
	Thread thisThread = Thread.currentThread();
	if (previous == null) {
		reEntranceDepth = 1;
		currentThread = thisThread;
	}
	else if (previous == thisThread) {
		reEntranceDepth++;
	}
	else {
		throw new IllegalStateException(
				"Concurrent access to KryoSerializer. Thread 1: " + thisThread.getName() +
						" , Thread 2: " + previous.getName());
	}
}

private void exitExclusiveThread() {
	if (--reEntranceDepth == 0) {
		currentThread = null;
	}
}
```